core/src/mast/merger.rs

```diff
@@ -0,0 +1,655 @@
+            // We need to remap the node prior to computing the EqHash.
+            //
+            // This is because the EqHash computation looks up its descendants and decorators in
+            // the internal index, and if we were to pass the original node to that
+            // computation, it would look up the incorrect descendants and decorators.
+            //
+            // Remapping at this point is going to be "complete", meaning all ids of children will
+            // be remapped since the DFS iteration guarantees that all children of this `node` have
+            // been processed before this node and their indices have been added to the
+            // mappings.
+            let remapped_node = self.remap_node(node, decorator_id_remapping, node_id_remapping);
+
+            let node_eq =
+                EqHash::from_mast_node(self.mast_forest, &self.hash_by_node_id, &remapped_node);
+
+            match self.node_id_by_hash.get(&node_eq) {
+                Some(existing_entry) => {
+                    // We have to map any occurence of `merging_id` to `existing_node_id`.
+                    node_id_remapping.insert(merging_id, *existing_entry);
+
+                    // Replace the external node in the existing forest with the non-external node
+                    // from the merging forest.
+                    // Any node in the existing forest that pointed to the external node will
+                    // have the same digest due to the semantics of external nodes.
+                    //
+                    // Note that the inverse case is handled implicitly, that is, an external node
+                    // in the merging forest that already exists as an external or non-external
+                    // node in the existing forest will not be added to the merged forest because it
+                    // will be handled like any other duplicate.
+                    if self.mast_forest[*existing_entry].is_external()
+                        && !remapped_node.is_external()
+                    {
+                        self.mast_forest[*existing_entry] = remapped_node;
+                    }
+                },
+                None => {
+                    let new_node_id = self.mast_forest.add_node(remapped_node)?;
+                    node_id_remapping.insert(merging_id, new_node_id);
+
+                    // We need to update the indices with the newly inserted nodes
+                    // since the EqHash computation requires all descendants of a node
+                    // to be in this index. Hence when we encounter a node in the merging forest
+                    // which has descendants (Call, Loop, Split, ...), then those need to be in the
+                    // indices.
+                    self.node_id_by_hash.insert(node_eq, new_node_id);
+                    self.hash_by_node_id.insert(new_node_id, node_eq);
+                },
+            }
```

Comment:
Question: let's say we have two forests where one has a procedure and the other has an external node referring to this procedure. Let's also say that `eq_hash` for these nodes do not match (e.g., because the full procedure has some decorators). Would this code still get rid of the external node?

Basically, I'm wondering if for external nodes we should check equality based on MAST roots, and if there is another non-external node with the same MAST root, we get rid of the external node.

Reply:
I don't think it would get rid of the external node in that case, as of right now. But indeed it sounds desirable to do this.
I'll add a test for this case and try to implement the desired behavior :+1:

Reply:
To merge external nodes with/without decorators and non-external nodes with/without decorators there are multiple options we can use.

## Case 1: External node has decorators

### Option 1:

The external node has decorators and its behavior-preserving replacement would be copying the referenced node and adding the external node's decorators.

```text
[External(foo, Trace(0))]
+
[Block(foo)]
=
[
  Block(foo)
  Block(foo, [Trace(0)]) <- replacement for the external node
]
```

### Option 2:

Decorators from the external node are **not** added to the referenced node's decorators, so its behavior changes. On the positive side we only have one node in the merged forest.

```text
[External(foo, Trace(0))]
+
[Block(foo)]
=
[
  Block(foo) <- replacement for the external node
]
```

## Case 2: Referenced node has decorators

### Option 1:

The external node has no decorators and its behavior-preserving replacement would be copying the referenced node except for its decorators.

```text
[External(foo)]
+
[Block(foo, Trace(0))]
=
[
  Block(foo) <- replacement for the external node
  Block(foo, [Trace(0)])
]
```

### Option 2:

The external node has no decorators but the node it references by MAST root has decorators, so they are merged.

```text
[External(foo)]
+
[Block(foo, Trace(0))]
=
[
  Block(foo, [Trace(0)]) <- replacement for the external node
]
```

## Case 3: Both nodes have decorators

### Option 1:

The external node has decorators and its behavior-preserving replacement would be copying the referenced node without decorators and adding the external node's decorators.

```text
[External(foo, Trace(0))]
+
[Block(foo, Trace(1))]
=
[
  Block(foo, Trace(1))
  Block(foo, Trace(0)) <- replacement for the external node
]
```

### Option 2:

Decorators from the external node are prepended (before_enter) and appended (after_exit) to the node they replace (i.e. decorators wrap around the existing decorators), creating a node with merged decorator behavior.

```text
[External(foo, Trace(0))]
+
[Block(foo, Trace(1))]
=
[
  Block(foo, Trace(1))
  Block(foo, Trace(0), Trace(1)) <- replacement for the external node
]
```

What you suggested is probably Case 2 / Option 2. For now I implemented behavior-preserving merging, but either one is possible.

Reply:
I'm thinking we could simply just drop the external node and not worry about its decorators - basically:

- Option 2 from case 1.
- Option 2 from case 2.
- For case 3, we'd just keep the reference node as it is (i.e., don't pre-pend anything). It would look something like this:

```
[External(foo, Trace(0))]
+
[Block(foo, Trace(1))]
=
[
  [Block(foo, Trace(1))]
]
```